lib/_http_parser.js

```diff
@@ -0,0 +1,890 @@
+          this.onBody(data, offset, nbytes);
+          offset += nbytes;
+          nbytes = 0;
+          this._state = STATE_BODY_CHUNKED_BYTES_CRLF;
+        } else {
+          nbytes -= dataleft;
+          this.onBody(data, offset, dataleft);
+          offset = len;
+        }
+        break;
+      case STATE_BODY_CHUNKED_BYTES_CRLF:
+        if (nbytes === 0 && data[offset++] === CR)
+          ++nbytes;
+        else if (nbytes === 1 && data[offset++] === LF)
+          this._state = STATE_BODY_CHUNKED_SIZE;
+        else
+          return this._setError('Malformed chunk (missing CRLF)');
+        break;
+      default:
+        return this._setError('Unexpected parser state while reading chunks');
+    }
+  }
+
+  this._buf = buf;
+  this._seenCR = seenCR;
+  this._nbytes = nbytes;
+
+  return len;
+};
+HTTPParser.prototype._executeBodyLiteral = function(data) {
+  var len = data.length;
+
+  if (len === 0)
+    return 0;
+
+  var nbytes = this._contentLen;
+  if (len >= nbytes) {
+    this.onBody(data, 0, nbytes);
+    this.onComplete && this.onComplete();
+    if ((this._flags & FLAG_SHOULD_KEEP_ALIVE) > 0) {
+      this.reinitialize(this.type);
+      if (len > nbytes) {
+        var ret = this.execute(data.slice(nbytes));
+        if (typeof ret !== 'number')
+          return ret;
+        return nbytes + ret;
+      }
+    } else
+      this._state = STATE_COMPLETE;
+  } else {
+    this._contentLen -= len;
+    this.onBody(data, 0, len);
+  }
+  return len;
+};
+HTTPParser.prototype._executeBodyEOF = function(data) {
+  var len = data.length;
+
+  if (len === 0)
+    return 0;
+
+  this.onBody(data, 0, len);
+  return len;
+};
+HTTPParser.prototype._executeBodyIgnore = function(data) {
+  return 0;
+};
+HTTPParser.prototype._executeError = function(data) {
+  return this._err;
+};
+HTTPParser.prototype.execute = HTTPParser.prototype._executeStartLine;
+HTTPParser.prototype._shouldKeepAlive = function(httpMajor, httpMinor, flags,
+                                                 type, status) {
+  if (httpMajor > 0 && httpMinor > 0) {
+    if ((flags & FLAG_CONNECTION_CLOSE) > 0)
+      return false;
+  } else {
+    if ((flags & FLAG_CONNECTION_KEEP_ALIVE) === 0)
+      return false;
+  }
+  return !this._needsEOF(flags, type, status);
+};
+HTTPParser.prototype._needsEOF = function(flags, type, status) {
+  if (type === REQUEST)
+    return false;
+
+  // See RFC 2616 section 4.4
+  if (status === 204 ||                    // No Content
+      status === 304 ||                    // Not Modified
+      parseInt(status / 100, 10) === 1 ||  // 1xx e.g. Continue
+      (flags & FLAG_SKIPBODY) > 0) {       // response to a HEAD request
+    return false;
+  }
+
+  if ((flags & FLAG_CHUNKED) > 0 || this._contentLen !== undefined)
+    return false;
+
+  return true;
+};
+
+
+
+
+var REQUEST = HTTPParser.REQUEST = 0;
+HTTPParser.RESPONSE = 1;
+module.exports = HTTPParser;
+
+function indexOfCRLF(buf, buflen, offset) {
```

Comment:
@dchusovitin As mentioned in my performance improvement techniques, the custom `indexOfCRLF()` is faster than `buffer.indexOf()`.